garage/tf/policies/maml_policy.py

```diff
@@ -0,0 +1,151 @@
+
+        self.wrapped_policy = wrapped_policy
+        self.n_tasks = n_tasks
+        self.name = name
+        self._initialized = False
+        self._adaptation_step_size = adaptation_step_size
+        self._adapted_param_store = dict()
+        self._create_update_opts()
+
+        super().__init__(wrapped_policy._env_spec)
+
+    def initialize(self, gradient_var, inputs):
+        """
+        Initialize the MAML Policy.
+
+        This funtion will create all the one step adapted parameters
+        and rebuild a model using these adapted parameters for each
+        task. A MamlPolicy can only be initialized once.
+
+        Args:
+            gradient_var: Gradient variables for one step adaptation.
+            inputs: Input tensors for rebuilding the models.
+
+        Returns:
+            all_model_outputs: The outputs of rebuilt models.
+            update_opts: Adapted parameter tensors.
+            update_opts_input: Input for calculating adapted parameters.
+        """
+
+        assert not self._initialized, "The MAML policy is initialized and can be initialized once."
+
+        global MAML_VARIABLE_STORE, MAML_TASK_INDEX
+        update_opts = []
+        with tf.name_scope(self.name):
+            # One step adaptation
+            with tf.name_scope("Adaptation"):
+                for i in range(self.n_tasks):
+                    params = self.wrapped_policy.get_params()
+                    gradient_i = gradient_var[i]
+
+                    for p, g in zip(params, gradient_i):
+                        adapted_param = p - self._adaptation_step_size * g
+                        name = "maml_policy/{}/{}".format(i, p.name)
+                        self._adapted_param_store[name] = adapted_param
+                        if i == 0:
+                            update_opts.append(adapted_param)
+
+            def maml_get_variable(name, shape=None, **kwargs):
+                scope = tf.get_variable_scope()
+                idx = 0
+                if MAML_TASK_INDEX >= self.n_tasks:
+                    raise ValueError("Invalid TASK_INDEX.")
+                fullname = "maml_policy/{}/{}/{}:{}".format(
+                    MAML_TASK_INDEX, scope.name, name, idx)
+                while fullname in MAML_VARIABLE_STORE:
+                    idx += 1
+                    fullname = "maml_policy/{}/{}/{}:{}".format(
+                        MAML_TASK_INDEX, scope.name, name, idx)
+                MAML_VARIABLE_STORE.add(fullname)
+                return self._adapted_param_store[fullname]
+
+            # build the model with these parameters
+            model = self.wrapped_policy.model
+
+            # overload the whole tf.get_variable function
+            # this allows us to use an operation as a variable
+            from tensorflow.python.ops import variable_scope
+            original_get_variable = variable_scope.get_variable
+            variable_scope.get_variable = maml_get_variable
```

Comment:
is there a reason you hot-patch this rather than using a context manager from `unittest.mock.patch`?

i'm worried this could cause strange (and difficult to find) bugs, especially if this was ever executed in multithreaded code. AFAIK this hot-patches get_variable process-wide